### ISSUE TITLE
Enhance Python interface for ease of use

### ISSUE THREAD

**Issue by russfellows** [russfellows/s3dlio#90]
# Review of AWS s3pytorchconnector vs. s3dlio
Here’s a code-level, side-by-side take on **awslabs/s3-connector-for-pytorch** (“AWS connector”) vs **russfellows/s3dlio** (“s3dlio”) with a focus on the **Python / PyTorch** interface and day-1 usability.

---
## Summary

* **Interface parity:** For core PyTorch use, **yes—s3dlio exposes substantially the same user-facing concepts**: iterable & map datasets, one-liner dataset construction from a URI/prefix, async loaders, and checkpoint read/write. Names and defaults differ, and the **item type you get back differs** (AWS returns an object with key+bucket+read(); s3dlio usually yields raw `bytes`). ([GitHub][1]) 
* **Ease of use:** **AWS is currently easier to pick up**: it’s `pip install` + very concise examples and polished docs; the `from_prefix(...)` factories feel extremely ergonomic. s3dlio is close in code ergonomics, adds multi-framework and multi-backend power, and now documents one-liner helpers (e.g., `create_async_loader`, `create_dataset`)—but **packaging & “batteries-included” examples aren’t quite as turnkey yet**. ([GitHub][1]) 

---

### AWS connector (awslabs/s3-connector-for-pytorch)

* **Datasets:** `S3IterableDataset`, `S3MapDataset` with clean **`.from_prefix(...)`** constructors. Iterable supports sharding, region, transform, etc. Map supports random access. Typical loop returns an **S3 object wrapper** where you can access **`.bucket`, `.key`, `.read()`**.
* **Minimal usage (from README):**

  ```py
  dataset = s3torchconnector.S3IterableDataset.from_prefix(
      bucket="compat.store", prefix="dlio_test_mini/azure", region="us-east-1",
      enable_sharding=True, shuffle=True, recursive=True, start_after="prefix/"
  )
  dataloader = DataLoader(dataset, batch_size=64, num_workers=4)
  for obj in dataloader:
      # obj has .bucket, .key, .read()
      ...
  ```

  (Exact pattern and options shown in README usage block.) ([GitHub][1])
* **Extras:** `S3Checkpoint` helpers (including Lightning integration); `S3StorageReader/Writer`; configuration via environment/kwargs; **install is `pip install s3torchconnector`**. ([GitHub][1])

### s3dlio

* **Universal helpers:** **`create_async_loader(uri, opts)`** and **`create_dataset(uri, opts)`**; plus `get/put/list/delete/stat` utilities. These are **backend-agnostic** (S3, Azure, local `file://`, etc.) and optimized for AI/ML throughput. 
* **Datasets:** `s3dlio.torch` exposes **`S3IterableDataset` and `S3MapDataset`** (re-exported in the Python package) so you can plug directly into `torch.utils.data.DataLoader`. Example patterns in tests/imports look like:

  ```py
  from s3dlio.torch import S3IterableDataset
  ds = S3IterableDataset("file:///some/dir", loader_opts={"file_pattern": "*.npz", "shuffle": True})
  for item in ds:
      ...
  ```

  (Class and import shown in repo; example test code iterates with `DataLoader`.)  
* **What you get per item:** the default async loader/dataset **yields `bytes`** (you decode/parse as needed). (Docs + examples explicitly show `async for item in loader: len(item)`.) 
* **Checkpointing:** **`create_checkpoint_writer(uri)` / `create_checkpoint_reader(uri)`** with streaming, zero-copy Python bindings in the code.  
* **Loader options:** rich realism knobs for training—shuffle, prefetch, workers, seeds, pin_memory, persistent_workers, timeout, channels_first/last, DDP sampling, etc. 
* **Install/dev flow:** active docs show building wheels (e.g., via build scripts) and troubleshooting; there are references to `pip install s3dlio`, but publishing status can vary—if not on PyPI, you build locally and `pip install` the wheel.  

---

## Are the **interfaces essentially the same** for Python (PyTorch) users?

**Mostly, yes for the core data-loading story**, with two notable differences:

1. **Construction style**

   * **AWS:** elegant **`Class.from_prefix(...)`** factories on `S3IterableDataset`/`S3MapDataset`, tightly tailored for S3. ([GitHub][1])
   * **s3dlio:** two main patterns

     * **Universal helpers** `create_dataset(uri, opts)` / `create_async_loader(uri, opts)` that autodetect the backend from the URI (e.g., `s3://...`, `file://...`). 
     * **Torch classes** `S3IterableDataset`/`S3MapDataset` available for direct use with `DataLoader`. 
       *Implication:* You’ll write a **different one-liner**, but it’s still a one-liner.

2. **What each item looks like**

   * **AWS** yields an **object with metadata** (`bucket`, `key`) and a **`.read()`** method; easy to pass into PIL/NumPy with an inline `transform`. ([GitHub][1])
   * **s3dlio** yields **raw `bytes`** by default (fast/zero-copy); you perform your own decode/parse—or enable a transform in your pipeline (loader options include `enable_transforms`).  
     *Implication:* If you depend on **key/bucket metadata inside the batch**, the AWS connector feels more ready out-of-the-box; s3dlio favors **speed & simplicity of bytes** unless you build a small wrapper.

Beyond that:

* **Both offer Iterable and Map datasets** for PyTorch. (AWS: documented classes; s3dlio: `s3dlio.torch` exports both.)  
* **Both support dataset sharding/parallelism knobs**; s3dlio’s `LoaderOptions` exposes many training-oriented toggles (pin_memory, persistent_workers, DDP sampling, etc.). ([GitHub][1]) 
* **Both implement checkpoint read/write** (AWS via `S3Checkpoint`; s3dlio via `create_checkpoint_*` and streaming bindings).   

**Conclusion on parity:** For day-to-day **PyTorch** loading, **s3dlio gives you practically the same capabilities with slightly different entry points and item shape**—plus the upside that the very same code paths can hit **S3, Azure, local `file://`, direct IO**, etc.

---

## Which is **easier** to use today?

* **AWS connector wins on “instant-on”.** It’s `pip install s3torchconnector`, then `S3IterableDataset.from_prefix(...)` and you’re training, with crisp API docs and examples, including Lightning helpers and checkpoint recipes. ([GitHub][1])
* **s3dlio is close in code ergonomics** (one-liner helpers, torch dataset classes), but **bytes-first returns** sometimes mean a tiny wrapper for image/tensor decode, and packaging/examples are still maturing. Docs show **universal helpers (`create_async_loader`, `create_dataset`)** and **clear troubleshooting**, but you may sometimes need to **build/install a wheel locally** if a PyPI artifact isn’t available in your environment.  

---

## Where s3dlio clearly **goes beyond** the AWS connector

* **Multi-framework:** APIs and tests span **PyTorch, TensorFlow, and JAX**. 
* **Multi-backend:** Works with **S3, Azure, `file://`, and direct I/O**, all via a single URI-driven API. 
* **Performance features:** **Zero-copy bytes**, async batching/prefetch, **adaptive tuning**, and rich `LoaderOptions` that mirror PyTorch training realities (pin_memory, persistent_workers, channels_last, DDP sampling, etc.).  

---

## Gaps / friction to close parity on “feel”

1. **Item object vs bytes.** Consider an optional **`return="record"`** mode (or a tiny `S3Record` dataclass) that yields `{bytes, key, bucket}` so users can drop in metadata-dependent transforms just like AWS. (AWS returns an object with `.bucket`, `.key`, `.read()`). ([GitHub][1])
2. **Factory ergonomics.** Add **`S3IterableDataset.from_prefix(...)`** and **`S3MapDataset.from_prefix(...)`** shorthands that forward to your universal constructors, keeping parity with AWS muscle memory.
3. **Packaging polish.** Ensure **published wheels** for common Python versions/architectures (and mention them prominently). Your docs currently show wheel builds & even reference `pip install s3dlio`, but users benefit from seeing the exact PyPI command as the canonical path. 
4. **Copy-paste snippets.** Mirror AWS’s README style: **5–10 line end-to-end** PyTorch examples (Iterable + Map, S3 + file://), **plus** a one-liner Lightning/Checkpoint example. You already have all the primitives (`create_checkpoint_*`). 
5. **Transforms.** Since you already have `enable_transforms` knobs, add a short **image decode** and **NPZ/TFRecord** sample (bytes → tensor) right under the quick start. 

---

## Bottom line

* **Do you provide essentially the same Python interface for PyTorch?**
  **Yes—functionally.** You cover Iterable/Map datasets, async loaders, sharding, and checkpointing. The key **differences are construction ergonomics** (AWS’s `from_prefix`) and the **item type** (AWS object wrapper with metadata vs s3dlio’s `bytes`).  
* **Is s3dlio as easy to use?**
  **Close, but AWS still feels more “plug-and-play” today** thanks to pip availability, copy-paste snippets, and polished docs. That said, s3dlio’s **universal URI API + multi-framework support** and **training-oriented loader options** are meaningful advantages once you’re past installation. ([GitHub][1]) 

## To Do
Create a **compat adapter** so `s3dlio` can optionally yield AWS-style items (`bucket`, `key`, `read()` that wraps the bytes) to ease migration—no changes to your core I/O needed.

[1]: https://github.com/awslabs/s3-connector-for-pytorch "GitHub - awslabs/s3-connector-for-pytorch: The Amazon S3 Connector for PyTorch delivers high throughput for PyTorch training jobs that access and store data in Amazon S3."
